nntrainer/models/dynamic_training_optimization.h

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file   dynamic_training_optimization.h
+ * @date   4 January 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is Dynamic Training Optimization for Neural Network
+ *
+ * Dynamic training aims to optimize the cost of applying the gradient.
+ * The cost of applying the gradient includes the cost of the optimizer (adam,
+ * etc) where the optimizer variables are updated, and the cost of actually
+ * updating the weights (which can be non-trivial with bigger weights and
+ * distributed training).
+ *
+ * There are two supported modes:
+ * 1. Gradient Mode: The already calculated gradient is used to estimate if this
+ * gradient must be used to update the weight, or if this update must be
+ * skipped.
+ *
+ * 2. Derivative Mode: This mode tries to estimate an approximate gradient with
+ * low cost in order to save the cost of calculating gradient. This cost of
+ * calculating gradient is wasted if the gradient is not going to be applied.
+ *
+ * There are two supported reduction operations which reduce the gradient and
+ * the weight to a single value in order to compare it with a threshold.
+ * If the reduced value is less than threshold, the update is performed with
+ * some probabilty proportional to the value. If the reduced value is higher
+ * than threshold, then the update is always performed.
+ *
+ */
+
+#ifndef __DYNAMIC_TRAINING_OPT_H__
+#define __DYNAMIC_TRAINING_OPT_H__
+#ifdef __cplusplus
+
+#include <random>
+#include <vector>
+
+#include <layer_internal.h>
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @class   DynamicTraining Optimization
+ * @brief   Dynamic Training Optimization
+ */
+class DynamicTrainingOptimization {
+public:
+  /**
+   * @brief     Constructor of DynamicFineTuning Optimization
+   */
+  DynamicTrainingOptimization(int threshold_ = 1, int skip_n_iter = 1);
+
+  /**
+   * @brief     Set threshold for optimization
+   */
+  void setThreshold(float threshold_) { threshold = threshold_; };
```

Comment:
need to add check if the threshold is greater than 1?

Reply:
Threshold is used a divider against the reduced value. 
So threshold can be anything but zero.

Reply:
I think at least we have to give them a notification what threshold which greater than 1.0 means. if the threshold is greater than 1, then it means only layer which approximation of gradient is greater than weight is updated. IMO, most of layer does not updated in normal cases.

Reply:
Then that notification should be give for even smaller thresholds.
Threshold of 1e-3 for resnet50 led to 75% skip of gradient while 1e-3 for VGG16 led to over 90% skip.
So, I am not sure what notification can be given for threshold over 1, as there is no direct meaning to the threshold values by itself.

Reply:
does it means most layers are going to update if threshold over 1 is used.  Am I right? If it is, then it is kind of the opposite of user's intention.

Reply:
So, the value of threshold cannot be directly compared to 1. Its range is a lot dependent on model, data and training configuration.

Reply:
> Its range is a lot dependent on model, data and training configuration.

Well, if you say so.. let's see. leave as it is.